realtek: add support for Panasonic Switch-M*eG PN28xx0K devices

musashino205 opened the pull request:
This PR contains some commits to add supports for the following 3x switches of Panasonic:

- Switch-M16eG PN28160K
- Switch-M24eG PN28240K
- Switch-M48eG PN28480K

These models has the RTL8218FB chip and it provides 2x or 4x combo port of TP/SFP, but the current drivers don't support it and only TP ports are available at the moment.

Note: I'm trying to add RTL8218FB support to networking-related drivers
https://github.com/musashino-build/openwrt/tree/devadd/m24eg_rtl8218fb

musashino205 committed:
realtek: add support for Panasonic Switch-M24eG PN28240K

Panasonic Switch-M24eG PN28240K is a 24 + 2 port gigabit switch, based on
RTL8382M.

Specification:

- SoC           : Realtek RTL8382M
- RAM           : DDR3 128 MiB (Winbond W631GG8KB-15)
- Flash         : SPI-NOR 32 MiB (Macronix MX25L25635FMI-10G)
- Ethernet      : 10/100/1000 Mbps x24 + 2
  - port  1-8   : TP, RTL8218B
  - port  9-16  : TP, RTL8218B (SoC)
  - port 17-24  : RTL8218FB
    - port 17-22: TP
    - port 23-24: TP/SFP (Combo)
- LEDs/Keys     : 7x / 1x
- UART          : RS-232 port on the front panel (connector: RJ-45)
  - 3:TX, 4:GND, 5:GND, 6:RX (pin number: RJ-45)
  - 9600n8
- Power         : 100-240 VAC, 50/60 Hz, 0.5 A
  - Plug        : IEC 60320-C13
- Stock OS      : VxWorks based

Flash instruction using initramfs image:

1.  Prepare the TFTP server with the IP address 192.168.1.111
2.  Rename the OpenWrt initramfs image to "0101A8C0.img" and place it to
    the TFTP directory
3.  Download the official upgrading firmware (ex: pn28240k_v30000.rom)
    and place it to the TFTP directory
4.  Boot M24eG and interrupt the U-Boot with Ctrl + C keys
5.  Execute the following commands and boot with the OpenWrt initramfs
    image

    rtk network on
    tftpboot 0x81000000
    bootm

6.  Backup mtdblock files to the computer by scp or anything and reboot
7.  Interrupt the U-Boot and execute the following commands to re-create
    filesystem in the flash

    ffsmount c:/
    ffsfmt c:/

    this step takes a long time, about ~ 4 mins

8.  Execute the following commands to put the official images to the
    filesystem

    updatert <official image>

    example:

      updatert pn28240k_v30000.rom

    this step takes about ~ 40 secs

9.  Set the environment variables of the U-Boot by the following commands

    setenv loadaddr 0xb4e00000
    setenv bootcmd bootm
    saveenv

10: Download the OpenWrt initramfs image and boot with it

    tftpboot 0x81000000 0101A8C0.img
    bootm

11: On the initramfs image, download the sysupgrade image and perform
    sysupgrade with it

    sysupgrade <imagename>

12: Wait ~ 120 seconds to complete flashing

Known Issues:

- 2x SFP ports are provided as combo ports by the RTL8218FB chip, but the
  phy driver has no support for it. Currently, only TP ports work by the
  RTL8218D support.

Note:

- "Switch-M24eG" is a model name, and "PN28240K" is a model number.
  Switch-M24eG has an another (old) model number ("PN28240"), it's not a
  Realtek based hardware.

- Switch-M24eG has a "POWER" LED (Green), but it's not connected to any
  GPIO pin.

- U-Boot checks the runtime images in the flash when booting and fails
  to execute "bootcmd" variable if the images are not existing.

- A filesystem is formed in the flash (0x100000-0x1DFFFFF) on the stock
  firmware and it includes the stock images, configuration files and
  checksum files. It's unknown format, can't be managed on the OpenWrt.
  To get the enough space for OpenWrt, move the filesystem to the head
  of "fs_reserved" partition by execution of "ffsfmt" and "updatert".

- A GPIO pin on PCA9539 is used for resetting external RTL8218B phy and
  RTL8218FB phy.
  This should be specified as "reset-gpios" property in MDIO node, but
  the current configuration of RTL8218B phy in the phy driver seems to
  be incomplete and RTL8218FB won't be configured on RTL8218D support.
  So, ethernet ports on these phys will be broken after hard-resetting.
  At the moment, configure this pin as gpio-hog to avoid breaking by
  resetting.

Back to the stock firmware:

1. Delete "loadaddr" variable and set "bootcmd" to the original value

   on U-Boot:

     setenv loadaddr
     setenv bootcmd 'bootm 0x81000000'

   on OpenWrt:

     fw_setenv loadaddr
     fw_setenv bootcmd 'bootm 0x81000000'

2. Perform reset or reboot

  on U-Boot:

    reset

  on OpenWrt:

    reboot

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: add support for Panasonic Switch-M16eG PN28160K

Panasonic Switch-M16eG PN28160K is a 16 + 2 port gigabit switch, based on
RTL8382M.

Specification:

- SoC           : Realtek RTL8382M
- RAM           : DDR3 128 MiB (Winbond W631GG8KB-15)
- Flash         : SPI-NOR 32 MiB (Macronix MX25L25635FMI-10G)
- Ethernet      : 10/100/1000 Mbps x16 + 2
  - port 1-8    : TP, RTL8218B (SoC)
  - port 9-16   : RTL8218FB
    - port  9-14: TP
    - port 15-16: TP/SFP (Combo)
- LEDs/Keys     : 7x / 1x
- UART          : RS-232 port on the front panel (connector: RJ-45)
  - 3:TX, 4:GND, 5:GND, 6:RX (pin number: RJ-45)
  - 9600n8
- Power         : 100-240 VAC, 50/60 Hz, 0.5 A
  - Plug        : IEC 60320-C13
- Stock OS      : VxWorks based

Flash instruction using initramfs image:

1.  Prepare the TFTP server with the IP address 192.168.1.111
2.  Rename the OpenWrt initramfs image to "0101A8C0.img" and place it to
    the TFTP directory
3.  Download the official upgrading firmware (ex: pn28160k_v30003.rom)
    and place it to the TFTP directory
4.  Boot M16eG and interrupt the U-Boot with Ctrl + C keys
5.  Execute the following commands and boot with the OpenWrt initramfs
    image

    rtk network on
    tftpboot 0x81000000
    bootm

6.  Backup mtdblock files to the computer by scp or anything and reboot
7.  Interrupt the U-Boot and execute the following commands to re-create
    filesystem in the flash

    ffsmount c:/
    ffsfmt c:/

    this step takes a long time, about ~ 4 mins

8.  Execute the following commands to put the official images to the
    filesystem

    updatert <official image>

    example:

      updatert pn28160k_v30003.rom

    this step takes about ~ 40 secs

9.  Set the environment variables of the U-Boot by the following commands

    setenv loadaddr 0xb4e00000
    setenv bootcmd bootm
    saveenv

10: Download the OpenWrt initramfs image and boot with it

    tftpboot 0x81000000 0101A8C0.img
    bootm

11: On the initramfs image, download the sysupgrade image and perform
    sysupgrade with it

    sysupgrade <imagename>

12: Wait ~ 120 seconds to complete flashing

Known Issues:

- 2x SFP ports are provided as combo ports by the RTL8218FB chip, but the
  phy driver has no support for it. Currently, only TP ports work by the
  RTL8218D support.

Note:

- "Switch-M16eG" is a model name, and "PN28160K" is a model number.
  Switch-M16eG has an another (old) model number ("PN28160"), it's not a
  Realtek based hardware.

- Switch-M16eG has a "POWER" LED (Green), but it's not connected to any
  GPIO pin.

- U-Boot checks the runtime images in the flash when booting and fails
  to execute "bootcmd" variable if the images are not existing.

- A filesystem is formed in the flash (0x100000-0x1DFFFFF) on the stock
  firmware and it includes the stock images, configuration files and
  checksum files. It's unknown format, can't be managed on the OpenWrt.
  To get the enough space for OpenWrt, move the filesystem to the head
  of "fs_reserved" partition by execution of "ffsfmt" and "updatert".

- A GPIO pin on PCA9539 is used for resetting external RTL8218FB phy.
  This should be specified as "reset-gpios" property in MDIO node, but
  RTL8218FB won't be configured on RTL8218D support in the phy driver.
  So, ethernet ports on the phy will be broken after hard-resetting.
  At the moment, configure this pin as gpio-hog to avoid breaking by
  resetting.

Back to the stock firmware:

1. Delete "loadaddr" variable and set "bootcmd" to the original value

   on U-Boot:

     setenv loadaddr
     setenv bootcmd 'bootm 0x81000000'

   on OpenWrt:

     fw_setenv loadaddr
     fw_setenv bootcmd 'bootm 0x81000000'

2. Perform reset or reboot

  on U-Boot:

    reset

  on OpenWrt:

    reboot

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: add support for Panasonic Switch-M48eG PN28480K

Panasonic Switch-M48eG PN28480K is a 48 + 4 port gigabit switch, based on
RTL8393M.

Specification:

- SoC           : Realtek RTL8393M
- RAM           : DDR3 128 MiB (Winbond W631GG8KB-15)
- Flash         : SPI-NOR 32 MiB (Macronix MX25L25635FMI-10G)
- Ethernet      : 10/100/1000 Mbps x48 + 2
  - port  1-40  : TP, RTL8218B x5
  - port 41-48  : RTL8218FB
    - port 41-44: TP
    - port 45-48: TP/SFP (Combo)
- LEDs/Keys     : 7x / 1x
- UART          : RS-232 port on the front panel (connector: RJ-45)
  - 3:TX, 4:GND, 5:GND, 6:RX (pin number: RJ-45)
  - 9600n8
- Power         : 100-240 VAC, 50/60 Hz, 0.5 A
  - Plug        : IEC 60320-C13
- Stock OS      : VxWorks based

Flash instruction using initramfs image:

1.  Prepare the TFTP server with the IP address 192.168.1.111
2.  Rename the OpenWrt initramfs image to "0101A8C0.img" and place it to
    the TFTP directory
3.  Download the official upgrading firmware (ex: pn28480k_v30000.rom)
    and place it to the TFTP directory
4.  Boot M48eG and interrupt the U-Boot with Ctrl + C keys
5.  Execute the following commands and boot with the OpenWrt initramfs
    image

    rtk network on
    tftpboot 0x81000000
    bootm

6.  Backup mtdblock files to the computer by scp or anything and reboot
7.  Interrupt the U-Boot and execute the following commands to re-create
    filesystem in the flash

    ffsmount c:/
    ffsfmt c:/

    this step takes a long time, about ~ 4 mins

8.  Execute the following commands to put the official images to the
    filesystem

    updatert <official image>

    example:

      updatert pn28480k_v30000.rom

    this step takes about ~ 40 secs

9.  Set the environment variables of the U-Boot by the following commands

    setenv loadaddr 0xb4e00000
    setenv bootcmd 'sleep 10; bootm;'
    saveenv

    'sleep 10;' is required as dummy to execute 'bootm' command correctly

10: Download the OpenWrt initramfs image and boot with it

    tftpboot 0x81000000 0101A8C0.img
    bootm

11: On the initramfs image, download the sysupgrade image and perform
    sysupgrade with it

    sysupgrade <imagename>

12: Wait ~ 120 seconds to complete flashing

Known Issues:

- 4x SFP ports are provided as combo ports by the RTL8218FB chip, but the
  phy driver has no support for it. Currently, only TP ports work by the
  RTL8218B support.

Note:

- "Switch-M48eG" is a model name, and "PN28480K" is a model number.
  Switch-M48eG has an another (old) model number ("PN28480"), it's not a
  Realtek based hardware.

- Switch-M48eG has a "POWER" LED (Green), but it's not connected to any
  GPIO pin.

- U-Boot checks the runtime images in the flash when booting and fails
  to execute "bootcmd" variable if the images are not existing.

- A filesystem is formed in the flash (0x100000-0x1DFFFFF) on the stock
  firmware and it includes the stock images, configuration files and
  checksum files. It's unknown format, can't be managed on the OpenWrt.
  To get the enough space for OpenWrt, move the filesystem to the head
  of "fs_reserved" partition by execution of "ffsfmt" and "updatert".

- A GPIO pin on PCA9539 is used for resetting external RTL8218B phys and
  RTL8218FB phy.
  This should be specified as "reset-gpios" property in MDIO node, but
  the current configuration of RTL8218B phy in the driver seems to be
  incomplete and RTL8218FB won't be configured on RTL8218D support.
  So, ethernet ports on these phys will be broken after hard-resetting.
  At the moment, configure this pin as gpio-hog to avoid breaking by
  resetting.

- This model has 2x Microchip TCN75A thermal sensors. Linux Kernel
  supports TCN75 chip on lm75 driver, but no support for TCN75'A'
  variant.
  At the moment, use TCN75 support for the chips instead.

Back to the stock firmware:

1. Delete "loadaddr" variable and set "bootcmd" to the original value

   on U-Boot:

     setenv loadaddr
     setenv bootcmd 'ffsrdm c:/runtime.had 0x81000000;alphadec c:/runtime.had 0x81000240 0x80010000;'

   on OpenWrt:

     fw_setenv loadaddr
     fw_setenv bootcmd 'ffsrdm c:/runtime.had 0x81000000;alphadec c:/runtime.had 0x81000240 0x80010000;'

2. Perform reset or reboot

  on U-Boot:

    reset

  on OpenWrt:

    reboot

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: enable pca953x GPIO driver for rtl839x subtarget

The system status LED on Panasonic Switch-M48eG PN28480K is connected to
a PCA9539PW. To use the LED as a status LED of OpenWrt while booting,
enable the pca953x driver and built-in to the kernel.
Also enable CONFIG_GPIO_PCA953X_IRQ to use interrupt via RTL83xx GPIO.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

svanheule commented:
As I've come to expect from your patches, they look quite good from a first quick look. I don't currently have much time, but I will try to give these patches a proper look soon. Especially the M48eG looks impressive; first time I see thermal zones being used here.

svanheule commented on target/linux/realtek/dts-5.10/rtl8382_panasonic_m16eg-pn28160k.dts:
```diff
@@ -0,0 +1,162 @@
+		gpios = <&gpio2 1 GPIO_ACTIVE_LOW>;
+		color = <LED_COLOR_ID_AMBER>;
+		function = LED_FUNCTION_STATUS;
+		function-enumerator = <1>;
```
You don't really need the function-enumerator when you already have two color to distinguish the LEDs by, as that's sufficient to prevent name collisions. But it's not incorrect to specify it either, so we can keep them. If you don't define label, the generated label would be "amber:status-1".

musashino205 replied:
Indeed, it's not required in this case. But the documentation of LED in the Linux Kernel says ```Integer to be used when more than one instance of the same function is needed, differing only with an ordinal number.```, so I would like to keep it.

svanheule replied:
Sure, the documentation doesn't disagree with you :)

svanheule commented on target/linux/realtek/dts-5.10/rtl8382_panasonic_m16eg-pn28160k.dts:
```diff
@@ -0,0 +1,162 @@
+		gpios = <&gpio2 2 GPIO_ACTIVE_LOW>;
+		color = <LED_COLOR_ID_GREEN>;
+		function = LED_FUNCTION_STATUS;
+		function-enumerator = <2>;
```
Would result in "green:status-2".

svanheule commented on target/linux/realtek/dts-5.10/rtl8382_panasonic_m16eg-pn28160k.dts:
```diff
@@ -0,0 +1,162 @@
+		gpio-hog;
+		gpios = <12 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "ext-switch-reset";
```
If this is the reset line for the RTL8218FB phy, shouldn't that go into `reset-gpios` on the MDIO (device) node?

musashino205 replied:
Indeed, it seems to be better than gpio-hog in this case.

musashino205 replied:
I tried to use of ```reset-gpios``` instead, but it breaks networking on Switch-M48eG PN28480K. The GPIO pin for resetting of phys will be requested before probing I2C and PCA9539 chip, then, the kernel shows some errors and networking will be broken.

```
[    5.632936] Probing RTL838X eth device pdev: 82111a00, dev: 82111a10
[    5.746637] Found SoC ID: 8393: RTL8393, family 8390
[    5.812299] Using MAC 0000005040101ab6
[    5.866193] mdio_bus mdio-bus: mii_bus mdio-bus couldn't get reset GPIO
[    5.954230] Error setting up netdev, freeing it again.
...
[    6.324806] pca953x 0-0075: using no AI
...
[    6.637545] RTL839X Chip-Info: a0016290, version B
[    6.701197] Probing RTL838X eth device pdev: 82111a00, dev: 82111a10
[    6.813966] Found SoC ID: 8393: RTL8393, family 8390
[    6.879439] Using MAC 0000005040101ab6
[    6.933084] mdio_bus mdio-bus: MDIO device at address 0 is missing.
[    8.270006] RTL839X Chip-Info: a0016290, version B
...
[   18.573278] In rtl83xx_vlan_setup
[   18.616963] UNKNOWN_MC_PMASK: 001fffffffffffff
[   18.675507] VLAN profile 0: L2 learning: 1, UNKN L2MC FLD PMSK 511,          UNKN IPMC FLD PMSK 511, UNKN IPv6MC FLD PMSK: 511
[   18.675518] VLAN profile 0: raw 000001ff, 003fe3ff
[   19.935092] rtl83xx_enable_phy_polling:     ffffffffffff
[   21.056181] rtl83xx-switch switch@1b000000 lan1 (uninitialized): PHY [rtl838x slave mii-0:00] driver [Realtek RTL8218B (external)] (irq=POLL)
[   21.227023] rtl83xx-switch switch@1b000000 lan2 (uninitialized): PHY [mdio-bus:01] driver [Realtek RTL8218B (external)] (irq=POLL)
[   21.385187] rtl83xx-switch switch@1b000000 lan3 (uninitialized): PHY [mdio-bus:02] driver [Realtek RTL8218B (external)] (irq=POLL)
[   21.543285] rtl83xx-switch switch@1b000000 lan4 (uninitialized): PHY [mdio-bus:03] driver [Realtek RTL8218B (external)] (irq=POLL)
...
```

change:
```
diff --git a/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts b/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts
index ca04c271eb..6f44a603f9 100644
--- a/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts
+++ b/target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts
@@ -179,14 +179,6 @@
        #interrupt-cells = <2>;
        interrupt-parent = <&gpio0>;
        interrupts = <18 IRQ_TYPE_EDGE_FALLING>;
-
-       /* GPIO12 (IO1_4): 5x RTL8218B + RTL8218FB */
-       ext_switch_reset {
-               gpio-hog;
-               gpios = <12 GPIO_ACTIVE_HIGH>;
-               output-high;
-               line-name = "ext-switch-reset";
-       };
 };

 &i2c_switch {
@@ -218,6 +210,7 @@
 &ethernet0 {
        mdio-bus {
                compatible = "realtek,rtl838x-mdio";
+               reset-gpios = <&gpio2 12 GPIO_ACTIVE_LOW>;
                #address-cells = <1>;
                #size-cells = <0>;

```

full log: https://gist.github.com/musashino205/372467aa470c4d8c89ced4f95e070499

svanheule replied:
Ideally the kernel would resolve device dependencies from DT, but I suppose that's not a trivial problem. What (probably) happens is that the GPIO layer returns `-EPROBE_DEFER` on the yet-unavailable line, causing the ethernet device to be probed again a bit later. The second time the GPIO device was available, so `reset-gpios` can be acquired as expected.

```
mdio_bus mdio-bus: MDIO device at address 0 is missing.
```
If just device@0 is missing, then maybe the MDIO subsystem is trying to probe the phy too soon after reset. Have you tried playing with the `reset-delay-us` and `reset-post-delay-us` values? Although I wonder why that would cause the ethernet to fail entirely, instead of just the first (eight) port(s).

musashino205 replied:
Hm, I tried ```reset-delay-us``` but ```reset-post-delay-us``` was not, I'll try.

musashino205 replied:
I tried ```reset-delay-us = <10000>``` and ```reset-post-delay-us = <10000>```, but still broken.

svanheule replied:
That's actually quite bad, because this is the way it's supposed to work. Can you tell what the delay is with `gpio-hog`, between the phy-s being taken out of reset and the driver probing the bus?

musashino205 replied:
BTW, the rtl83xx-phy driver doesn't configure RTL8218B phy on RTL839x platform, so the phys may not have to be hard-reset. (on RTL838x, ```rtl8380_configure_ext_rtl8218b``` will be called but not on RTL839x)

svanheule replied:
Ah, that's why this is breaking then. Can you add `reset-gpios` only on the phy-s where it does work? Otherwise you'll have to keep the hog with an explanation of why it needs to be done in this way. I would also suggest you create an issue ticket so this problem (RTL8218B reset on RTL839x) is documented.

musashino205 replied:
Hmm, this GPIO pin resets all phys on the MDIO line without internal phy on RTL838x SoC, so can't be used for the specified phys only.
And, RTL8218FB phy on **Switch-M16eG PN28160K** (RTL8382M) and **Switch-M24eG PN28240K** (RTL8382M) will be detected as RTL8218D by the phy driver on RTL838x platform, but RTL8218D won't be configured by the driver. So, ```reset-gpios``` may also not be used on that 2x models. I'll test ```reset-gpios``` on **Switch-M24eG PN28240K**.

musashino205 replied:
Oops, I tested on **Switch-M24eG PN28240K**, but in addition to RTL8218FB (detected as RTL8218D), external RTL8218B phy is also broken... (internal RTL8218B phy on RTL8382M SoC is good)
The configuration by the phy driver for RTL8218B may be imcomplete for hard-reset.

So, I would like to keep ```gpio-hog``` on 3x models in this PR.

svanheule replied:
Of course using it on one phy doesn't work if it's physically connected to all phy-s. Silly me :facepalm:  

Then `gpio-hog` with some explanation sounds like an okay way to document this in the meantime.

svanheule commented on target/linux/realtek/dts-5.10/rtl8382_panasonic_m16eg-pn28160k.dts:
```diff
@@ -0,0 +1,162 @@
+&ethernet0 {
+	mdio-bus {
+		compatible = "realtek,rtl838x-mdio";
+		regmap = <&ethernet0>;
```
I'm not sure if I noticed this before, and now I can't find any reference to "regmap" in `target/linux/realtek/files-5.10/`, nor in the upstream mdio binding. Do you know how this property is used?

musashino205 replied:
After checking, it seems that it existed from at least "[2b88563ee5 realtek: update the tree to the latest refactored version](https://github.com/openwrt/openwrt/commit/2b88563ee5aafd9571d965b7f2093a0f58d98a31)". I was taking it over by inertia, but it certainly looks like it has nothing to do and is unnecessary.

svanheule replied:
OK, then you should be able to leave it out here, and it can be removed from the other DT files to prevent more copy-pasting. To be clear, I'm not asking you to do that, you can just focus on these patches.

svanheule reviewed (commented):
Only one real remark, about the MDIO device reset. Looks good otherwise.

svanheule reviewed (commented):
Commit message has
```
- The U-Boot checks the runtime images in the flash when booting and
```
'U-Boot' or 'The bootloader'
```
  fails to execute anything in "bootcmd" variable if the images are not
```
'the "bootcmd" variable' or just 'fails to execute the "bootcmd" variable...'
```
  exsisting.
```
 and "exsisting" should be "existing". You probably have the same bit in the other commit messages. Otherwise this looks identical to the 16-port model (aside from the 8 extra ports), so the same remarks apply.

svanheule reviewed (approved):
Looks good to me.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts:
```diff
@@ -0,0 +1,352 @@
+	sda-gpios = <&gpio0 17 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+
+	/* Microchip TCN75A (for SoC) */
+	tsens0: sensor@48 {
```
Maybe change the label to something along the lines of "tsens_soc". Then this would be more descriptive, and you don't really need the comment anymore, unless you want to give the reader an idea of where on the board this sensor is.

musashino205 replied:
Sounds great, I'll update the label. However, the lm75 driver has no support for TCN75'A' variant and I would like to keep the comment at the moment.

svanheule replied:
I had to look it up, but the difference is apparently that the TCN75A has 12-bit measurement resolution, but the TCN75 only has 9-bit. Then it's OK to keep this comment. Could you add this info to the commit message too?

musashino205 replied:
> Could you add this info to the commit message too?

OK, I'll do.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts:
```diff
@@ -0,0 +1,352 @@
+		 *
+		 * Fan Speed setting:
+		 * - Low : 0-40 celsius
+		 * - High: 0-50 celsius
```
"40-50"?

musashino205 replied:
Switch-M48eG PN28480K provides 2x modes of fan speed, high/low. On the stock firmware, the speed need to be configured manually by user (default: high) and it won't be changed automatically depending on the measured temperature.
And the official recommended environment temperature depends on the setting,
- Low: 0-40 celsius
- High: 0-50 celsius

(official document (EN): https://panasonic.co.jp/ew/pewnw/ns/mno/eng/pdf/pn28480k-en_exp.pdf)

This time, I experimentally configured to change automatically depending on the measured temperatures on 2x TCN75A.

svanheule replied:
So you're actually improving thermal management. :)

Maybe clarify the comment that "Low" and "High" refer to the ambient temperature range, and are not the fan speeds for certain device temperatures.

musashino205 replied:
> Maybe clarify the comment that "Low" and "High" refer to the ambient temperature range, and are not the fan speeds for certain device temperatures.

OK, I'll update.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts:
```diff
@@ -0,0 +1,352 @@
+				map {
+					trip = <&sys_alert>;
+					cooling-device = <&fan 0 1>;
+				};
```
How does the stock firmware deal with the two fans? Are they also run independently, or do they turn on/off together? I'm wondering how much sense it makes, physically, to consider this as two independent thermal zones if the fans are actually mounted in the chassis, and not e.g. on the CPU.

musashino205 replied:
2x fans of Switch-M48eG PN28480K are controlled integrally like 1x fan.
The thermal zone of Linux Kernel supports only one sensor per zone, so 2x zones are required for 2x sensors, SoC and system. Both zones controls the same fan unit.

I checked behavior on the device (set thresholds soc->37, sys->36 for test):

- SoC: 35.5 celsius, Sys: 32.5 celsius ---> speed: Low

- SoC: 37.5 celsius, Sys: 34.5 celsius ---> speed: High

- SoC: 40.5 celsius, Sys: 36.0 celsius ---> speed: High

- SoC: 37.0 celsius, Sys: 33.5 celsius ---> speed: High

- SoC: 36.5 celsius, Sys: 33.0 celsius ---> speed: Low

svanheule replied:
Right, there is only one fan section. My mind must have interpreted the two speeds as two fans... My apologies!

All very good then. Having the both zones independently enable the fans if either gets too hot sounds exactly like what one would want.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts:
```diff
@@ -0,0 +1,352 @@
+	};
+
+	/* Microchip TCN75A (for System) */
+	tsens1: sensor@49 {
```
This one could be labelled "tsens_system".

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts:
```diff
@@ -0,0 +1,352 @@
+	interrupt-parent = <&gpio0>;
+	interrupts = <18 IRQ_TYPE_EDGE_FALLING>;
+
+	/* GPIO12 (IO1_4): 5x RTL8218B + RTL8218FB */
```
If one GPIO line resets all phy-s this should definitely go into `reset-gpios` on the MDIO bus node.

svanheule reviewed (commented):
Since the rest is similar to the other devices, mainly remarks about the thermal management here.

svanheule commented:
The four reviews are for the respective patches. Apparently Github doesn't let you give one global comment on all different commits, but still insists that the comments apply to all changes in the PR :roll_eyes:

robimarko commented on target/linux/realtek/dts-5.10/rtl8382_panasonic_m16eg-pn28160k.dts:
```diff
@@ -0,0 +1,162 @@
+	/*
+	 * sfp0/1 are "combo" port with each TP port (23/24), and they are
+	 * connected to the RTL8218FB. Currently, there is no support for
+	 * the chip and only TP ports work by the RTL8218D support.
```
I guess "Twisted Pair", so RJ45 copper

svanheule replied:
That would make sense. I figured it must apply to the copper ports, but hadn't made the connection yet.

musashino205 commented on target/linux/realtek/dts-5.10/rtl8393_panasonic_m48eg-pn28480k.dts:
```diff
@@ -0,0 +1,352 @@
+		compatible = "gpio-fan";
+		gpios = <&gpio0 19 GPIO_ACTIVE_HIGH>;
+		gpio-fan,speed-map = <6000 0
+				      8900 1>;
```
Oh, okay, I'll update.

musashino205 commented:
Thank you for your review, I'm checking/updating/testing suggested changes. And also, I'll correct the commit messages you pointed out.

musashino205 commented:
UPDATE:

- ALL
  - drop ```regmap``` property from mdio node (M16eG, M24eG, M48eG)
  - update comment of ```ext_switch_reset``` gpio-hog (M16eG, M24eG, M48eG)
  - update commit messages
    - corrected syntax and words
    - added some notes

- Switch-M48eG
  - use descriptive DT label on TCN75A (M48eG)
  - add comment and update fan settings in gpio-fan (M48eG)
  - update comment of thermal-zones (M48eG)

svanheule commented:
Merged. Thanks!

3d669ec9cd14 realtek: add support for Panasonic Switch-M16eG PN28160K
fa52e5e25428 realtek: add support for Panasonic Switch-M24eG PN28240K
4974ee7341eb realtek: enable pca953x GPIO driver for rtl839x subtarget
cef3f05a55a3 realtek: add support for Panasonic Switch-M48eG PN28480K
